src/transaction.js

```diff
@@ -327,37 +408,92 @@ Transaction.prototype.byteLength = function () {
+  var byteLength = 0
+  byteLength += varuint.encodingLength(this.vShieldedOutput.length)  // nShieldedOutput
+  byteLength += (948 * this.vShieldedOutput.length)  // vShieldedOutput
+  return byteLength
+}
+
+Transaction.prototype.getJoinSplitByteLength = function () {
+  // TODO: Throw if we are using the wrong network
+  if (!coins.isZcash(this.network)) {
     return 0
   }
 
-  if (!coins.isZcash(this.network)) {
+  if (this.version < Transaction.ZCASH_JOINSPLITS_SUPPORT_VERSION) {
     return 0
   }
 
   var pubkeySigLength = (this.joinsplits.length > 0) ? (32 + 64) : 0
+
   return (
     bufferutils.varIntSize(this.joinsplits.length) +
     this.joinsplits.reduce(function (sum, joinsplit) {
-      return (
-        sum +
-        8 + 8 + 32 +
-        joinsplit.nullifiers.length * 32 +
-        joinsplit.commitments.length * 32 +
-        32 + 32 +
-        joinsplit.macs.length * 32 +
-        65 + 33 * 7 +
-        joinsplit.ciphertexts.length * Transaction.ZCASH_NOTECIPHERTEXT_SIZE
-      )
+      var byteLength = 0
+      byteLength += 8  // vpub_old
+      byteLength += 8  // vpub_new
+      byteLength += 32  // anchor
+      byteLength += joinsplit.nullifiers.length * 32  // nullifiers
+      byteLength += joinsplit.commitments.length * 32  // commitments
+      byteLength += 32  // ephemeralKey
+      byteLength += 32  // randomSeed
+      byteLength += joinsplit.macs.length * 32  // vmacs
+      if (this.version >= Transaction.ZCASH_SAPLING_VERSION) {
+        byteLength += 192  // zkproof Groth16 encoding
+      } else {
+        byteLength += 296  // zkproof PHGR13 encoding
+      }
+      byteLength += joinsplit.ciphertexts.length * Transaction.ZCASH_NOTECIPHERTEXT_SIZE
+      return sum + byteLength
     }, 0) +
     pubkeySigLength
   )
 }
 
+Transaction.prototype.zcashTransactionByteLength = function () {
+  var byteLength = 0
+  byteLength += 4  // Header
+  if (this.version >= Transaction.ZCASH_OVERWINTER_VERSION) {
+    byteLength += 4  // nVersionGroupId
+  }
+  byteLength += varuint.encodingLength(this.ins.length)  // tx_in_count
+  byteLength += this.ins.reduce(function (sum, input) { return sum + 40 + varSliceSize(input.script) }, 0)  // tx_in
+  byteLength += varuint.encodingLength(this.outs.length)  // tx_out_count
+  byteLength += this.outs.reduce(function (sum, output) { return sum + 8 + varSliceSize(output.script) }, 0)  // tx_out
+  byteLength += 4  // tx_out
+  if (this.version >= Transaction.ZCASH_OVERWINTER_VERSION) {
+    byteLength += 4  // nExpiryHeight
+  }
+  if (this.version >= Transaction.ZCASH_SAPLING_VERSION) {
+    byteLength += 8  // valueBalance
+    byteLength += this.getShieldedSpendByteLength()
+    byteLength += this.getShieldedOutputByteLength()
+  }
+  if (this.version >= Transaction.ZCASH_JOINSPLITS_SUPPORT_VERSION) {
+    byteLength += this.getJoinSplitByteLength()
+  }
+  if (this.version >= Transaction.ZCASH_SAPLING_VERSION &&
```

Comment:
readability: you make this same check 3 times in this file, refactor it into its own function and give it a nice name

Reply:
Please resolve this once it's refactored.

Reply:
Done